### PR TITLE
feat(ddl): allow alter table with generated columns 

### PR DESCRIPTION
**Pull request by yuhao-su**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- allow alter table with generated columns
- if the column to drop is referenced by a generated column, RW will yield error. User can proceed drop after dropping the generated column

TODO
- We need allow dropping generated column in table with schema registry. Currently not allowing drop any column from it.

#8756

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

We can now alter a table with generated columns. If we mentioned in the doc we can't, we need to update it.

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/frontend/src/handler/alter_table_with_sr.rs [hunk 1]
```diff
@@ -81,14 +75,34 @@ pub async fn handle_refresh_schema(
...
+    match result {
+        Ok(_) => Ok(RwPgResponse::empty_result(StatementType::ALTER_TABLE)),
+        Err(e) => {
+            let report = e.to_report_string();
+            // This is a workaround for reporting errors when columns to drop is referenced by generated column.
+            // Finding the actual columns to drop requires generating `PbSource` from the sql definition
+            // and fetching schema from schema registry, which will cause a lot of unnecessary refactor.
+            // The only cause of those matched errors is the generated column referencing the column to drop.
+            if report.contains("Failed to bind expression")
+                && report.contains("Item not found: Invalid column")
```
Comment (yuhao-su) [inline src/frontend/src/handler/alter_table_with_sr.rs:96]: Hope you like it @BugenZhao

File: e2e_test/source_inline/kafka/avro/alter_table.slt [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+Caused by:
+  Permission denied: columns to drop is referenced by generated columns: Failed to bind expression: bar + 1: Item not found: Invalid column: bar
+
+
+# TODO: Add support for dropping generated columns for table with schema registry
+statement error
+ALTER TABLE t DROP COLUMN gen_col;
+----
+db error: ERROR: Failed to run the query
```
Comment (yuhao-su) [inline e2e_test/source_inline/kafka/avro/alter_table.slt:58]: We need allow dropping generated column in table with schema registry. Currently not allowing drop any column from it.

File: src/frontend/src/handler/alter_table_column.rs [hunk 3]
```diff
@@ -193,6 +187,24 @@ pub async fn handle_alter_table_column(
...
 
+            // Check if the column to drop is referenced by any generated columns.
+            for column in original_catalog.columns() {
+                if let Some(expr) = column.generated_expr() {
+                    let expr = ExprImpl::from_expr_proto(expr)?;
+                    let refs = expr.collect_input_refs(original_catalog.columns().len());
+                    for idx in refs.ones() {
+                        let refed_column = &original_catalog.columns()[idx];
+                        if refed_column.name() == column_name.real_value() {
+                            Err(ErrorCode::PermissionDenied(format!(
```
Comment (BugenZhao) [inline src/frontend/src/handler/alter_table_column.rs:198]: Permission denied does not sound to be an appropriate category. 🤣

File: src/frontend/src/handler/alter_table_with_sr.rs [hunk 4]
```diff
@@ -81,14 +75,34 @@ pub async fn handle_refresh_schema(
...
+            {
+                Err(ErrorCode::PermissionDenied(format!(
+                    "columns to drop is referenced by generated columns: {}",
```
Comment (BugenZhao) [inline src/frontend/src/handler/alter_table_with_sr.rs:99]: Can we unify the error message?

File: src/frontend/src/handler/alter_table_with_sr.rs (hunk 1 above)
Comment (BugenZhao) [reply src/frontend/src/handler/alter_table_with_sr.rs:96]: It's obscure and unintuitive that "failed to bind expression" is specific to binding the expression for generated columns here. I would rather suggest adding a context like "failed to bind generated column" below, then exactly match that prompt.

https://github.com/risingwavelabs/risingwave/blob/3142219c0411713cf37ff1cdacf2c8ee802674da/src/frontend/src/handler/create_table.rs#L291

File: src/frontend/src/handler/alter_table_column.rs (hunk 3 above)
Comment (yuhao-su) [reply src/frontend/src/handler/alter_table_column.rs:198]: I don't like it either. But we are using it everywhere https://github.com/singularity-data/risingwave/blob/739d0059f3b7a76975d503821549fd808fef8af9/src/meta/src/manager/catalog/mod.rs#L592

File: src/frontend/src/handler/alter_table_with_sr.rs (hunk 4 above)
Comment (yuhao-su) [reply src/frontend/src/handler/alter_table_with_sr.rs:99]: No. Here we don't know which column is being dropped

File: src/frontend/src/handler/alter_table_with_sr.rs [hunk 5]
```diff
@@ -81,14 +76,35 @@ pub async fn handle_refresh_schema(
...
+            // Here we match the error message to yield when failing to bind generated column exprs.
+            let re = Regex::new(r#"fail to bind expression in generated column "(.*?)""#).unwrap();
+            let captures = re.captures(&report).map_err(anyhow::Error::from)?;
+            if let Some(gen_col_name) = captures.and_then(|captures| captures.get(1)) {
+                Err(ErrorCode::PermissionDenied(format!(
+                    "columns to drop are referenced by a generated column \"{}\": {}",
+                    gen_col_name.as_str(),
+                    report,
+                ))
+                .into())
```
Comment (BugenZhao) [inline src/frontend/src/handler/alter_table_with_sr.rs:104]:
```rust
anyhow!(e).context("columns to drop ..")
```

**Review by BugenZhao — COMMENTED** [review]
Generally LGTM as it just works

**Review by BugenZhao — APPROVED** [review]
not bad

**Pull request merged**